[clang][CodeGen] Set `dead_on_return` when passing arguments indirectly

antoniofrighetto opened the pull request:
Let Clang emit `dead_on_return` attribute on pointer arguments that are passed indirectly, namely, large aggregates that the ABI mandates be passed by value; thus, the parameter is destroyed within the callee. Writes to such arguments are not observable by the caller after the callee returns.

This should desirably enable further MemCpyOpt/DSE optimizations.

Previous discussion: https://discourse.llvm.org/t/rfc-add-dead-on-return-attribute/86871.

antoniofrighetto commented:
Experiencing some ASan failures, I suspect we are adding the attribute on arguments where we shouldn't (possibly C++ destructors).

antoniofrighetto commented:
Updated, thanks, not sure how I overlooked that; hopefully should be on track.

rjmccall commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -2852,8 +2852,17 @@ void CodeGenModule::ConstructAttributeList(StringRef Name,
+        const auto *RD = ParamType->getAsCXXRecordDecl();
+        if (!RD || (RD && RD->hasTrivialDestructor()))
+          Attrs.addAttribute(llvm::Attribute::DeadOnReturn);
+      }
```
Please use:

```
  if (!ParamType.isDestructedType() || !ParamType->isRecordType() ||
      ParamType->castAs<RecordType>()->getDecl()->isParamDestroyedInCallee())
```

We do destroy parameters in the callee in several situations. The most important example is the MSVC C++ ABI, but I believe that coincidentally never uses indirect argument passing, so it wouldn't be testable. The testable example would be something like this in Objective-C++:

```
struct HasWeakReference {
  __weak id ref;
};
void test(HasWeakReference) {} // should still be dead_on_return because the caller is not responsible to destroy it
```

`__weak` helpfully forces the struct to be passed indirectly, but you could also just use a `__strong id ref;` in a struct that's big enough to be passed indirectly on the target.

antoniofrighetto replied:
Thank you for clarifying this, updated it. Tests updated too (it indeed looks like from the tests that some MSVC callee-destroyed objects, which are passed indirectly, should be dead_on_return. So is the case for Obj-C++ structs, as mentioned, with __weak and __strong references, as per `objc-struct-cxx-abi.mm` and `ptrauth-struct-cxx-abi.mm` tests).

rjmccall reviewed (commented):
Thanks. I haven't done a detailed review of all the test changes, but the functionality now looks good to me. Your comment is subtly misleading, though; you were thinking about C++ when you wrote it, but it applies more generally. LGTM other than that.

antoniofrighetto commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -2852,8 +2852,21 @@ void CodeGenModule::ConstructAttributeList(StringRef Name,
+        // Add dead_on_return when the object's lifetime ends in the callee.
+        // This includes trivially-destructible C++ objects, as well as objects
+        // whose destruction / clean-up is carried out within the callee (e.g.,
+        // Obj-C++ ARC-managed structs, MSVC callee-destroyed objects).
```
Oh, sure thing, thanks!

rjmccall reviewed (approved):
Thanks, that LGTM. Looks like you have a merge conflict to resolve, but otherwise this is ready to go.

antoniofrighetto committed:
[clang][CodeGen] Set `dead_on_return` when passing arguments indirectly

Let Clang emit `dead_on_return` attribute on pointer arguments
that are passed indirectly, namely, large aggregates that the
ABI mandates be passed by value; thus, the parameter is destroyed
within the callee. Writes to such arguments are not observable by
the caller after the callee returns.

This should desirably enable further MemCpyOpt/DSE optimizations.

Previous discussion: https://discourse.llvm.org/t/rfc-add-dead-on-return-attribute/86871.
